### PR TITLE
Tagged Durable Objects Migrations

### LINKED ISSUE #1950: wrangler publish --rename-class and --transfer-class arguments help is insufficient
#### Describe the bug

This is the help being displayed:

        --rename-class <rename-class> <rename-class>                           Rename a durable object class
        --transfer-class <transfer-class> <transfer-class> <transfer-class>
            Transfer all durable objects associated with a class in another script to a class in this script

#### Reproduce the bug

```
wrangler publish -h
```

#### Expected behavior

Should display something like

        --rename-class <OldClassName> <NewClassName>                     Rename a durable object class
        --transfer-class <old-script> <ClassNameInOldScript> <ClassNameInThisScript>
            Transfer all durable objects associated with a class in old-script to a class in this script

#### Environment and versions

* output of `wrangler -V`:

        wrangler 1.17.0

### PR DESCRIPTION
**Pull request by xortive**
fixes #1950 

This PR adds support for tagged migrations to wrangler. Migration tags are a way to ensure that you are applying the correct migration to a given script that contains a DO.

They are also used for allowing multiple migrations to be specified in a list in `wrangler.toml` -- wrangler will fetch the script's current migration tag and only send migrations after it in the list.

Migration tags can be used in two ways:

- Specifying `--old-tag` and `--new-tag` on `wrangler publish`.
  `--old-tag` can only be omitted if the script doesn't have a migration tag already, otherwise it must be provided (so it can be verified against the script's actual current migration tag)
  
- The `[[migrations]]` list in wrangler.toml

```toml
[durable_objects]
bindings = [{name = "COUNTER", class_name = "Counter"}]

[[migrations]]
tag = "v1"
new_classes = ["Counter"]
```

With `[[migrations]]` you can just `wrangler publish` without having to use `--new-class` on the first upload, wrangler will look at the current migration tag of the script and handle everything for you.

testing instructions:
* clone the `malonso/tagged-migrations` branch
* in the root of the repository run `cargo install --path . --debug`.
* add your migrations to `wrangler.toml`

full syntax for `[[migrations]]`
```
[[migrations]]
tag = "v1" # should be unique
new_classes = ["Counter"] # array of new classes
deleted_classes = ["DeprecatedClass"] # array of deleted classes
renamed_classes = [{from: "CounterOld", to: "CounterNew"}] # array of rename directives
transferred_classes = [{from_script: "oldCounterScript", from: "Counter", to: "Counter"}] # array of transfer directives


# The migrations list is ordered -- only migrations *after* the script's current migration tag will be run
[[migrations]]
tag = "v2"
# More directives here. You can have multiple types of directives in a single migration, with multiple instances of each directive.
```

### CODE REVIEW

File: src/settings/toml/migrations.rs [hunk 1]
```diff
@@ -1,44 +1,175 @@
...
-                old_tag: None,
+    fn validate(&self) -> Result<(), anyhow::Error> {
+        // validate migration tags
+        match &self {
+            Migrations::Adhoc {
+                script_tag: MigrationTag::HasTag(script_tag),
+                provided_old_tag: Some(provided_tag),
+                ..
+            } if provided_tag != script_tag => anyhow::bail!(
+                "The migration tag provided with your adhoc migration (\"{}\") does not match the script's current migration tag of \"{}\", \
```
Comment (a-robinson) [inline src/settings/toml/migrations.rs:43]: nit, but do users know what "adhoc" means in this context? I'd leave that word out.

Or if we leave it in, "ad hoc" is the much more common way of spelling it :)

File: src/settings/toml/migrations.rs [hunk 2]
```diff
@@ -1,44 +1,175 @@
...
+            } => {
+                let api_migration = ApiMigration {
+                    old_tag: provided_old_tag.to_owned(),
+                    new_tag: new_tag.to_owned(),
+                    steps: migration
+                        .as_ref()
+                        .map_or_else(Vec::new, |m| vec![m.clone()]),
+                };
+
+                log::info!("Api Migration: {:#?}", api_migration);
```
Comment (a-robinson) [inline src/settings/toml/migrations.rs:106]: Same question about whether this was intentional. If it was, I'd capitalize API.

Ditto below in the ::List case as well.

**Review by a-robinson — APPROVED** [review]
Looks great!

File: src/commands/publish.rs [hunk 3]
```diff
@@ -216,3 +228,39 @@ fn validate_target_required_fields_present(target: &Target) -> Result<()> {
...
+            ..
+        }) => MigrationTag::HasTag(tag),
+        Some(ScriptResponse {
+            migration_tag: None,
+            ..
+        }) => MigrationTag::NoTag,
+        None => MigrationTag::NoScript,
+    };
+
+    log::info!("Current MigrationTag: {:#?}", tag);
```
[the remark this answers is not in the document]
Comment (xortive) [reply src/commands/publish.rs:252]: these only show up with `RUST_LOG` set -- it's nice to have around for issues that get filed

File: src/commands/publish.rs [hunk 4]
```diff
@@ -216,3 +228,39 @@ fn validate_target_required_fields_present(target: &Target) -> Result<()> {
...
     Ok(())
 }
+
+fn get_migration_tag(client: &Client, target: &Target) -> Result<MigrationTag, anyhow::Error> {
+    let addr = format!(
+        "https://api.cloudflare.com/client/v4/accounts/{}/workers/scripts",
+        target.account_id.load()?
+    );
+
+    let res: ListScriptsV4ApiResponse = client.get(&addr).send()?.json()?;
```
[the remark this answers is not in the document]
Comment (xortive) [reply src/commands/publish.rs:238]: Good point! I'll add one

**Comment by nilslice** [comment]
Let's hold on merging this - @Electroid should review, and at least one dev prod eng team member. We're planning a release for tomorrow, and unsure if this should be included or not.

**Review by Electroid — APPROVED** [review]

**Comment by signalnerve** [comment]
can we make sure to start work on a docs PR when this goes out? thanks!!

**Pull request merged**